posts/2022-10-28-gats-stabilization.md

```diff
@@ -0,0 +1,275 @@
+
+```rust
+fn iterate<T, I: for<'a> LendingIterator<Item<'a> = &'a T>>(iter: I) {
+    iter.for_each(|_: &T| {})
+}
+```
+
+```rust
+error: `I` does not live long enough
+   |
+   |     iter.for_each(|_: &T| {})
+   |                   ^^^^^^^^^^
+```
+
+Well, that isn't great. Turns out, this is pretty closely related to the first limitation we talked about earlier. But has strong connections to the borrow checker.
+
+But, let's look at something that's very clearly a borrow checker problem, by looking at an implementation of the `Filter` struct returned by the `filter` method:
+
+```rust
+impl<I: LendingIterator, P> LendingIterator for Filter<I, P>
+where
+    P: FnMut(&I::Item<'_>) -> bool, // <- the bound from above, a function
+{
+    type Item<'a> = I::Item<'a> where Self: 'a; // <- Use the underlying type
+
+    fn next(&mut self) -> Option<I::Item<'_>> {
+        // Loop through each item in the underlying `LendingIterator`...
+        while let Some(item) = self.iter.next() {
+            // ...check if the predicate holds for the item...
+            if (self.predicate)(&item) {
+                // ...and return it if it does
+                return Some(item);
+            }
+        }
+        // Return `None` when we're out of items
+        return None;
+    }
+}
+```
+
+Again, the implementation here shouldn't seem surprising. We, of course, run into a borrow checker error:
+
+```rust
+error[E0499]: cannot borrow `self.iter` as mutable more than once at a time
+  --> src/main.rs:28:32
+   |
+27 |     fn next(&mut self) -> Option<I::Item<'_>> {
+   |             - let's call the lifetime of this reference `'1`
+28 |         while let Some(item) = self.iter.next() {
+   |                                ^^^^^^^^^^^^^^^^ `self.iter` was mutably borrowed here in the previous iteration of the loop
+29 |             if (self.predicate)(&item) {
+30 |                 return Some(item);
+   |                        ---------- returning this value requires that `self.iter` is borrowed for `'1`
+```
+
+This is a known limitation in the current borrow checker and should be solved in some future iteration (like [Polonius](https://github.com/rust-lang/polonius)).
+
+### Non-local requirements for where clauses on GATs
+
+The last limitation we'll talk about today is a bit different than the others; it's not a bug and it shouldn't prevent any programs from compiling. But it all comes back to that `where Self: 'a` clause you've seen in several parts of this post. As mentioned before, if you're interested in digging a bit into why that clause is required, see the [push for stabilization post][stabilization_post].
+
+There is one not-so-ideal about this clause: you must write it on the trait. Like with where clauses on functions, you cannot add clauses to associated types in impls that aren't there in the trait. However, if you *didn't* add this clause, a large set of potential impls of the trait would be disallowed.
```

Comment:
> There is one not-so-ideal <...?> about this clause

It feels like there is a word missing here, or that it's confusing if it's intended to be an adjective for an implied "limitation".